File: modules/mofitness-pose/ios/MofitnessPoseView.swift
import ExpoModulesCore
import Foundation
import UIKit

final class MofitnessPoseView: ExpoView {
  let onPoseFrame = EventDispatcher()
  let onStatusChange = EventDispatcher()

  private let previewView = UIView()
  private var cameraService: MofitnessPoseCameraService?
  private var landmarkerService: MofitnessPoseLandmarkerService?
  private let stateQueue = DispatchQueue(label: "com.mofitness.pose.ios.state")
  private var isActive = false
  private var isMounted = false
  private var cameraFacing = "front"
  private var requestedTargetFps = 30
  private var effectiveTargetFps = 30
  private var numPoses = 1
  private var minPoseDetectionConfidence: Float = 0.55
  private var minPosePresenceConfidence: Float = 0.55
  private var minTrackingConfidence: Float = 0.55
  private var inferenceInFlight = false
  private var lastSubmittedTimestampMs = 0
  private var processedFrames = 0
  private var droppedFrames = 0
  private var actualFps = 0
  private var lastStatsTimestampMs = 0
  private var processedFramesAtLastStats = 0
  private var thermalLevel = "nominal"
  private var currentStatus = "idle"
  private var lastStatusDispatchTimestampMs = 0
  private var lastDispatchedStatus = ""

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    backgroundColor = .black

    previewView.backgroundColor = .black
    addSubview(previewView)

    refreshThermalState()
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(handleThermalStateDidChange),
      name: ProcessInfo.thermalStateDidChangeNotification,
      object: nil
    )
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    previewView.frame = bounds
    cameraService?.updatePreviewFrame(previewView.bounds)
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    isMounted = window != nil
    updateStreamingState()
  }

  func setActive(_ active: Bool) {
    isActive = active
    if active {
      resetTelemetry()
    }
    updateStreamingState()
  }

  func setCameraFacing(_ facing: String) {
    cameraFacing = facing
    cameraService?.setCameraPosition(facing == "back" ? .back : .front)
  }

  func setTargetFps(_ fps: Int) {
    requestedTargetFps = min(max(fps, 10), 30)
    refreshThermalState()
  }

  func setNumPoses(_ poses: Int) {
    numPoses = min(max(poses, 1), 4)
    rebuildLandmarker()
  }

  func setMinPoseDetectionConfidence(_ confidence: Float) {
    minPoseDetectionConfidence = min(max(confidence, 0.1), 0.99)
    rebuildLandmarker()
  }

  func setMinPosePresenceConfidence(_ confidence: Float) {
    minPosePresenceConfidence = min(max(confidence, 0.1), 0.99)
    rebuildLandmarker()
  }

  func setMinTrackingConfidence(_ confidence: Float) {
    minTrackingConfidence = min(max(confidence, 0.1), 0.99)
    rebuildLandmarker()
  }

  @objc private func handleThermalStateDidChange() {
    refreshThermalState()
    emitStatus(currentStatus, message: nil, force: true)
  }

  private func updateStreamingState() {
    guard isMounted else {
      cameraService?.stop()
      return
    }

    if !isActive {
      cameraService?.stop()
      resetFrameState()
      emitStatus("paused", message: nil, force: true)
      return
    }

    emitStatus("initializing", message: nil, force: true)
    ensureServices()
    cameraService?.start()
  }

  private func ensureServices() {
    if cameraService == nil {
      let service = MofitnessPoseCameraService(previewView: previewView)
      service.delegate = self
      service.setCameraPosition(cameraFacing == "back" ? .back : .front)
      cameraService = service
    }

    if landmarkerService == nil {
      rebuildLandmarker()
    }
  }

  private func rebuildLandmarker() {
    landmarkerService = MofitnessPoseLandmarkerService(
      numPoses: numPoses,
      minPoseDetectionConfidence: minPoseDetectionConfidence,
      minPosePresenceConfidence: minPosePresenceConfidence,
      minTrackingConfidence: minTrackingConfidence,
      liveStreamDelegate: self
    )

    if landmarkerService == nil && isActive {
      emitStatus("error", message: "Failed to initialize the iOS pose landmarker.", force: true)
    }
  }

  private func refreshThermalState() {
    thermalLevel = Self.mapThermalState(ProcessInfo.processInfo.thermalState)
    effectiveTargetFps = Self.computeEffectiveTargetFps(baseTargetFps: requestedTargetFps, thermalLevel: thermalLevel)
  }

  private func resetTelemetry() {
    stateQueue.sync {
      inferenceInFlight = false
      lastSubmittedTimestampMs = 0
      processedFrames = 0
      droppedFrames = 0
      actualFps = 0
      lastStatsTimestampMs = 0
      processedFramesAtLastStats = 0
      lastStatusDispatchTimestampMs = 0
      lastDispatchedStatus = ""
      currentStatus = "idle"
    }
  }

  private func resetFrameState() {
    stateQueue.sync {
      inferenceInFlight = false
      lastSubmittedTimestampMs = 0
    }
  }

  private func recordProcessedFrame() {
    stateQueue.sync {
      processedFrames += 1
      let now = Int(Date().timeIntervalSince1970 * 1000)
      if lastStatsTimestampMs == 0 {
        lastStatsTimestampMs = now
        processedFramesAtLastStats = processedFrames
        return
      }

      let elapsedMs = now - lastStatsTimestampMs
      if elapsedMs < 1000 {
        return
      }

      let processedSinceLastStats = processedFrames - processedFramesAtLastStats
      actualFps = Int((Double(processedSinceLastStats) * 1000.0 / Double(elapsedMs)).rounded())
      lastStatsTimestampMs = now
      processedFramesAtLastStats = processedFrames
    }
  }

  private func emitStatus(_ status: String, message: String?, force: Bool = false) {
    let payload: [String: Any]? = stateQueue.sync {
      currentStatus = status
      let now = Int(Date().timeIntervalSince1970 * 1000)
      let shouldDispatch = force || status != lastDispatchedStatus || now - lastStatusDispatchTimestampMs >= 1000
      if !shouldDispatch {
        return nil
      }

      lastDispatchedStatus = status
      lastStatusDispatchTimestampMs = now

      var nextPayload: [String: Any] = [
        "status": status,
        "fps": effectiveTargetFps,
        "targetFps": effectiveTargetFps,
        "actualFps": actualFps,
        "droppedFrames": droppedFrames,
        "processedFrames": processedFrames,
        "thermalLevel": thermalLevel
      ]
      if let message, !message.isEmpty {
        nextPayload["message"] = message
      }
      return nextPayload
    }

    guard let payload else {
      return
    }

    DispatchQueue.main.async {
      self.onStatusChange(payload)
    }
  }

  private static func computeEffectiveTargetFps(baseTargetFps: Int, thermalLevel: String) -> Int {
    switch thermalLevel {
    case "moderate":
      return min(baseTargetFps, 20)
    case "serious", "critical", "shutdown":
      return min(baseTargetFps, 10)
    default:
      return baseTargetFps
    }
  }

  private static func mapThermalState(_ state: ProcessInfo.ThermalState) -> String {
    switch state {
    case .nominal:
      return "nominal"
    case .fair:
      return "fair"
    case .serious:
      return "serious"
    case .critical:
      return "critical"
    @unknown default:
      return "unknown"
    }
  }
}

extension MofitnessPoseView: MofitnessPoseCameraServiceDelegate {
  func cameraService(_ service: MofitnessPoseCameraService, didOutput sampleBuffer: CMSampleBuffer, orientation: UIImage.Orientation) {
    guard isActive, let landmarkerService else {
      return
    }

    let now = Int(Date().timeIntervalSince1970 * 1000)
    let minimumFrameIntervalMs = max(1, Int(1000 / max(effectiveTargetFps, 1)))
    let shouldSubmit = stateQueue.sync { () -> Bool in
      if inferenceInFlight || now - lastSubmittedTimestampMs < minimumFrameIntervalMs {
        droppedFrames += 1
        return false
      }

      inferenceInFlight = true
      lastSubmittedTimestampMs = now
      return true
    }

    if !shouldSubmit {
      return
    }

    landmarkerService.detectAsync(sampleBuffer: sampleBuffer, orientation: orientation, timestampInMilliseconds: now)
  }

  func cameraService(_ service: MofitnessPoseCameraService, didChangeStatus status: MofitnessPoseCameraStatus, message: String?) {
    switch status {
    case .ready:
      emitStatus("ready", message: message, force: true)
    case .paused:
      emitStatus("paused", message: message, force: true)
    case .permissionDenied:
      emitStatus("error", message: message ?? "Camera permission denied.", force: true)
    case .failed:
      emitStatus("error", message: message ?? "Camera startup failed.", force: true)
    }
  }
}

extension MofitnessPoseView: MofitnessPoseLandmarkerLiveStreamDelegate {
  func poseLandmarkerService(_ service: MofitnessPoseLandmarkerService, didFinishDetection result: MofitnessPoseResultBundle?, error: Error?) {
    let submittedTimestamp = stateQueue.sync { () -> Int in
      inferenceInFlight = false
      return lastSubmittedTimestampMs
    }

    if let error {
      emitStatus("error", message: error.localizedDescription, force: true)
      return
    }

    guard
      let result,
      let poseLandmarkerResult = result.poseLandmarkerResult,
      let firstPose = poseLandmarkerResult.landmarks.first,
      !firstPose.isEmpty
    else {
      emitStatus("no_pose", message: nil)
      return
    }

    recordProcessedFrame()
    let resolution = cameraService?.videoResolution ?? .zero
    let payloadLandmarks = firstPose.map { landmark in
      [
        "x": landmark.x,
        "y": landmark.y,
        "z": landmark.z,
        "visibility": 1.0
      ]
    }

    DispatchQueue.main.async {
      self.onPoseFrame([
        "landmarks": payloadLandmarks,
        "timestamp": submittedTimestamp,
        "frameWidth": Int(resolution.width),
        "frameHeight": Int(resolution.height),
        "inferenceTimeMs": result.inferenceTimeMs
      ])
      self.emitStatus("tracking", message: nil)
    }
  }
}

File: modules/mofitness-pose/android/src/main/java/expo/modules/mofitnesspose/MofitnessPoseView.kt
package expo.modules.mofitnesspose

import android.content.Context
import android.os.Build
import android.os.PowerManager
import android.os.SystemClock
import android.util.Size
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.roundToInt

class MofitnessPoseView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext),
  PoseLandmarkerStreamHelper.Listener {
  private val onPoseFrame by EventDispatcher()
  private val onStatusChange by EventDispatcher()
  private val previewView = PreviewView(context).apply {
    layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    implementationMode = PreviewView.ImplementationMode.PERFORMANCE
    scaleType = PreviewView.ScaleType.FILL_CENTER
  }
  private val analysisExecutor: ExecutorService = Executors.newSingleThreadExecutor()
  private val inferenceInFlight = AtomicBoolean(false)
  private val powerManager = context.getSystemService(Context.POWER_SERVICE) as? PowerManager
  private val analysisResolution = Size(1280, 720)
  private val thermalListener =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      PowerManager.OnThermalStatusChangedListener { status ->
        handleThermalStatusChanged(status)
      }
    } else {
      null
    }

  private var processCameraProvider: ProcessCameraProvider? = null
  private var landmarkerHelper: PoseLandmarkerStreamHelper? = null
  private var active = false
  private var lensFacing = CameraSelector.LENS_FACING_FRONT
  private var requestedTargetFps = 30
  private var effectiveTargetFps = 30
  private var numPoses = 1
  private var minPoseDetectionConfidence = 0.55f
  private var minPosePresenceConfidence = 0.55f
  private var minTrackingConfidence = 0.55f
  private var thermalLevel = "nominal"
  private var lastSubmittedAtMs = 0L
  private var processedFrames = 0L
  private var droppedFrames = 0L
  private var actualFps = 0
  private var lastStatsAtMs = 0L
  private var processedFramesAtLastStats = 0L
  private var currentStatus = "idle"
  private var lastDispatchedStatus = ""
  private var lastStatusDispatchAtMs = 0L
  private var thermalListenerRegistered = false

  init {
    addView(previewView)
    refreshThermalState()
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    registerThermalListenerIfNeeded()
    refreshThermalState()
    updateStreamingState()
  }

  override fun onDetachedFromWindow() {
    unregisterThermalListenerIfNeeded()
    stopStreaming()
    analysisExecutor.shutdownNow()
    super.onDetachedFromWindow()
  }

  fun setActive(value: Boolean) {
    if (active == value) {
      return
    }

    active = value
    if (active) {
      resetTelemetry()
    }
    updateStreamingState()
  }

  fun setCameraFacing(value: String) {
    val nextFacing = if (value == "back") CameraSelector.LENS_FACING_BACK else CameraSelector.LENS_FACING_FRONT
    if (lensFacing == nextFacing) {
      return
    }

    lensFacing = nextFacing
    rebindCamera()
  }

  fun setTargetFps(value: Int) {
    requestedTargetFps = value.coerceIn(10, 30)
    val previousEffectiveTarget = effectiveTargetFps
    refreshThermalState()
    if (active && previousEffectiveTarget != effectiveTargetFps) {
      rebindCamera()
    }
  }

  fun setNumPoses(value: Int) {
    numPoses = value.coerceIn(1, 4)
    recreateLandmarker()
  }

  fun setMinPoseDetectionConfidence(value: Float) {
    minPoseDetectionConfidence = value.coerceIn(0.1f, 0.99f)
    recreateLandmarker()
  }

  fun setMinPosePresenceConfidence(value: Float) {
    minPosePresenceConfidence = value.coerceIn(0.1f, 0.99f)
    recreateLandmarker()
  }

  fun setMinTrackingConfidence(value: Float) {
    minTrackingConfidence = value.coerceIn(0.1f, 0.99f)
    recreateLandmarker()
  }

  override fun onError(message: String) {
    inferenceInFlight.set(false)
    emitStatus("error", message = message, force = true)
  }

  override fun onResults(result: PoseLandmarkerStreamHelper.ResultBundle) {
    inferenceInFlight.set(false)
    recordProcessedFrame()
    val firstPose = result.result.landmarks().firstOrNull()

    if (firstPose.isNullOrEmpty()) {
      emitStatus("no_pose")
      return
    }

    val landmarks = firstPose.map { landmark ->
      mapOf(
        "x" to landmark.x().toDouble(),
        "y" to landmark.y().toDouble(),
        "z" to landmark.z().toDouble(),
        "visibility" to landmark.visibility().orElse(0f).toDouble(),
      )
    }

    onPoseFrame(
      mapOf(
        "landmarks" to landmarks,
        "timestamp" to result.result.timestampMs(),
        "frameWidth" to result.inputImageWidth,
        "frameHeight" to result.inputImageHeight,
        "inferenceTimeMs" to result.inferenceTimeMs,
      ),
    )
    emitStatus("tracking")
  }

  private fun updateStreamingState() {
    if (!active) {
      stopStreaming()
      emitStatus("paused", force = true)
      return
    }

    val activity = appContext.currentActivity
    if (activity !is LifecycleOwner) {
      emitStatus("unavailable", message = "A lifecycle-aware activity is required for native pose tracking.", force = true)
      return
    }

    emitStatus("initializing", force = true)
    ensureLandmarker()

    val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
    cameraProviderFuture.addListener(
      {
        try {
          val provider = cameraProviderFuture.get()
          processCameraProvider = provider
          bindCamera(provider, activity)
        } catch (error: Exception) {
          emitStatus("error", message = error.message ?: "Failed to start the camera pipeline.", force = true)
        }
      },
      ContextCompat.getMainExecutor(context),
    )
  }

  private fun ensureLandmarker() {
    if (landmarkerHelper != null) {
      return
    }

    analysisExecutor.execute {
      try {
        landmarkerHelper = PoseLandmarkerStreamHelper(
          context = context,
          listener = this,
          numPoses = numPoses,
          minPoseDetectionConfidence = minPoseDetectionConfidence,
          minPosePresenceConfidence = minPosePresenceConfidence,
          minTrackingConfidence = minTrackingConfidence,
        )
      } catch (error: Exception) {
        emitStatus("error", message = error.message ?: "Failed to initialize the pose landmarker.", force = true)
      }
    }
  }

  private fun recreateLandmarker() {
    analysisExecutor.execute {
      try {
        landmarkerHelper?.clear()
        landmarkerHelper = PoseLandmarkerStreamHelper(
          context = context,
          listener = this,
          numPoses = numPoses,
          minPoseDetectionConfidence = minPoseDetectionConfidence,
          minPosePresenceConfidence = minPosePresenceConfidence,
          minTrackingConfidence = minTrackingConfidence,
        )
      } catch (error: Exception) {
        emitStatus("error", message = error.message ?: "Failed to reconfigure the pose landmarker.", force = true)
      }
    }
  }

  private fun bindCamera(provider: ProcessCameraProvider, lifecycleOwner: LifecycleOwner) {
    val preview = Preview.Builder()
      .setTargetResolution(analysisResolution)
      .build()
      .also { it.setSurfaceProvider(previewView.surfaceProvider) }
    val imageAnalysis = ImageAnalysis.Builder()
      .setTargetResolution(analysisResolution)
      .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
      .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
      .build()

    imageAnalysis.setAnalyzer(analysisExecutor) { imageProxy ->
      if (!active) {
        imageProxy.close()
        return@setAnalyzer
      }

      val now = SystemClock.uptimeMillis()
      val minFrameIntervalMs = (1000L / effectiveTargetFps.coerceAtLeast(1)).coerceAtLeast(1L)
      if (inferenceInFlight.get() || now - lastSubmittedAtMs < minFrameIntervalMs) {
        droppedFrames += 1
        imageProxy.close()
        return@setAnalyzer
      }

      lastSubmittedAtMs = now
      inferenceInFlight.set(true)
      val helper = landmarkerHelper
      if (helper == null) {
        inferenceInFlight.set(false)
        imageProxy.close()
        emitStatus("initializing")
        return@setAnalyzer
      }

      helper.detectLiveStream(imageProxy, lensFacing == CameraSelector.LENS_FACING_FRONT)
    }

    try {
      provider.unbindAll()
      provider.bindToLifecycle(
        lifecycleOwner,
        CameraSelector.Builder().requireLensFacing(lensFacing).build(),
        preview,
        imageAnalysis,
      )
      emitStatus("ready", force = true)
    } catch (error: Exception) {
      emitStatus("error", message = error.message ?: "Failed to bind the camera use cases.", force = true)
    }
  }

  private fun rebindCamera() {
    if (!active) {
      return
    }
    updateStreamingState()
  }

  private fun stopStreaming() {
    processCameraProvider?.unbindAll()
    inferenceInFlight.set(false)
    lastSubmittedAtMs = 0L
    currentStatus = "idle"
  }

  private fun registerThermalListenerIfNeeded() {
    if (thermalListenerRegistered || Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      return
    }

    val listener = thermalListener ?: return
    powerManager?.addThermalStatusListener(listener)
    thermalListenerRegistered = true
  }

  private fun unregisterThermalListenerIfNeeded() {
    if (!thermalListenerRegistered || Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      return
    }

    val listener = thermalListener ?: return
    powerManager?.removeThermalStatusListener(listener)
    thermalListenerRegistered = false
  }

  private fun handleThermalStatusChanged(status: Int) {
    val previousTarget = effectiveTargetFps
    thermalLevel = mapThermalStatus(status)
    effectiveTargetFps = computeEffectiveTargetFps(requestedTargetFps, thermalLevel)

    if (active && previousTarget != effectiveTargetFps) {
      rebindCamera()
      return
    }

    emitStatus(currentStatus.ifBlank { "ready" }, force = true)
  }

  private fun refreshThermalState() {
    thermalLevel =
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        mapThermalStatus(powerManager?.currentThermalStatus ?: PowerManager.THERMAL_STATUS_NONE)
      } else {
        "nominal"
      }
    effectiveTargetFps = computeEffectiveTargetFps(requestedTargetFps, thermalLevel)
  }

  private fun computeEffectiveTargetFps(baseTargetFps: Int, level: String): Int {
    return when (level) {
      "moderate" -> minOf(baseTargetFps, 20)
      "serious", "critical", "shutdown" -> minOf(baseTargetFps, 10)
      else -> baseTargetFps
    }
  }

  private fun mapThermalStatus(status: Int): String {
    return when (status) {
      PowerManager.THERMAL_STATUS_NONE -> "nominal"
      PowerManager.THERMAL_STATUS_LIGHT -> "fair"
      PowerManager.THERMAL_STATUS_MODERATE -> "moderate"
      PowerManager.THERMAL_STATUS_SEVERE -> "serious"
      PowerManager.THERMAL_STATUS_CRITICAL,
      PowerManager.THERMAL_STATUS_EMERGENCY -> "critical"
      PowerManager.THERMAL_STATUS_SHUTDOWN -> "shutdown"
      else -> "unknown"
    }
  }

  private fun resetTelemetry() {
    droppedFrames = 0L
    processedFrames = 0L
    actualFps = 0
    lastStatsAtMs = 0L
    processedFramesAtLastStats = 0L
    lastDispatchedStatus = ""
    lastStatusDispatchAtMs = 0L
  }

  private fun recordProcessedFrame() {
    processedFrames += 1
    val now = SystemClock.uptimeMillis()
    if (lastStatsAtMs == 0L) {
      lastStatsAtMs = now
      processedFramesAtLastStats = processedFrames
      return
    }

    val elapsedMs = now - lastStatsAtMs
    if (elapsedMs < 1000L) {
      return
    }

    val processedSinceLastStats = processedFrames - processedFramesAtLastStats
    actualFps = ((processedSinceLastStats * 1000.0) / elapsedMs.toDouble()).roundToInt()
    lastStatsAtMs = now
    processedFramesAtLastStats = processedFrames
  }

  private fun emitStatus(status: String, message: String? = null, force: Boolean = false) {
    currentStatus = status
    val now = SystemClock.uptimeMillis()
    val shouldDispatch = force || status != lastDispatchedStatus || now - lastStatusDispatchAtMs >= STATUS_DISPATCH_INTERVAL_MS
    if (!shouldDispatch) {
      return
    }

    lastDispatchedStatus = status
    lastStatusDispatchAtMs = now

    val payload = mutableMapOf<String, Any>(
      "status" to status,
      "fps" to effectiveTargetFps,
      "targetFps" to effectiveTargetFps,
      "actualFps" to actualFps,
      "droppedFrames" to droppedFrames,
      "processedFrames" to processedFrames,
      "thermalLevel" to thermalLevel,
    )
    if (!message.isNullOrBlank()) {
      payload["message"] = message
    }
    onStatusChange(payload)
  }

  companion object {
    private const val STATUS_DISPATCH_INTERVAL_MS = 1000L
  }
}
